runners/spark/src/main/java/org/apache/beam/runners/spark/io/MicrobatchSource.java

```diff
@@ -201,56 +201,66 @@ public int hashCode() {
    */
```

Comment:
They definitely look clone-ish. A brief look indicates that the internal stuff in `BoundedReadFromUnboundedSource` are quite private so extending them in order perform the Spark-specific tweaks may require exposing these internals at the very least. 
Perhaps that was the reason it was cloned in the first place.

I'll try to take a deeper look and see what it would take to consolidate them.